File: backend/src/schema/paper.py
from typing import Optional
from pydantic import BaseModel
from datetime import datetime


class PaperCreate(BaseModel):
    title: str
    abstract: str
    authors: str
    arxiv_id: Optional[str] = None
    pdf_url: Optional[str] = None
    paper_url: Optional[str] = None
    github_url: Optional[str] = None
    topics: Optional[str] = None
    published_date: Optional[str] = None
    institution: Optional[str] = None
    date_published: Optional[str] = None
    thumbnail_url: Optional[str] = None
    paper_source: Optional[str] = "arxiv"


class PaperUploadMetadata(BaseModel):
    """Metadata for uploaded PDF papers (non-arxiv)"""

    title: str
    abstract: str
    authors: str
    # Optional fields
    github_url: Optional[str] = None
    topics: Optional[str] = None
    published_date: Optional[str] = None
    institution: Optional[str] = None
    date_published: Optional[str] = None


class PaperResponse(BaseModel):
    id: int
    user_id: int
    title: str
    abstract: str
    authors: str
    arxiv_id: Optional[str] = None
    pdf_url: Optional[str] = None
    paper_url: Optional[str] = None
    github_url: Optional[str] = None
    topics: Optional[str] = None
    published_date: Optional[str] = None
    thumbnail_url: Optional[str] = None
    institution: Optional[str] = None
    date_published: Optional[str] = None
    created_at: datetime
    ingested: bool
    paper_summary: Optional[str] = None
    paper_source: Optional[str] = "arxiv"
    keywords: Optional[str] = None


class BulkDeletePapers(BaseModel):
    paper_ids: list[int]


class RecommendationItem(BaseModel):
    arxiv_id: Optional[str] = None
    title: Optional[str] = None
    authors: Optional[str] = None
    categories: Optional[str] = None
    primary_category: Optional[str] = None
    published_date: Optional[str] = None
    paper_url: Optional[str] = None
    pdf_url: Optional[str] = None
    score: Optional[float] = None
    final_score: Optional[float] = None


class RecommendationsResponse(BaseModel):
    similar_papers: list[RecommendationItem] = []
    on_this_topic: list[RecommendationItem] = []
    from_these_authors: list[RecommendationItem] = []

File: frontend/src/types/paper.ts
export interface Paper {
  id: string;
  title: string;
  authors: string[];
  abstract: string;
  date: string;
  topics: string[];
  institution?: string;
  country?: string;
  pdfUrl: string;
  htmlUrl?: string;
  githubUrl?: string;
  thumbnailUrl?: string;
  primaryCategory?: string;
  ingested?: boolean;
  keywords?: string[];
}

export interface SavedPaper {
  id: number;
  user_id: number;
  title: string;
  abstract: string;
  authors: string;
  arxiv_id: string;
  pdf_url: string;
  paper_url?: string;
  github_url?: string;
  topics?: string;
  published_date?: string;
  thumbnail_url?: string;
  institution?: string;
  date_published?: string;
  created_at: string;
  ingested: boolean;
  keywords?: string | null;
}

export interface FilterOption {
  id: string;
  label: string;
  value?: string;
  count?: number;
  color: 'coral' | 'teal' | 'amber' | 'violet' | 'emerald' | 'blue' | 'rose';
}

export interface Filters {
  topics: string[];
  countries?: string[];
  institutions?: string[];
  years: string[];
}

export type ViewMode = 'grid' | 'detailed';

export interface RecommendationItem {
  arxiv_id: string | null;
  title: string | null;
  abstract: string | null;
  authors: string | null;
  categories: string | null;
  primary_category: string | null;
  published_date: string | null;
  paper_url: string | null;
  pdf_url: string | null;
  score: number | null;
  final_score: number | null;
}

export interface RecommendationsResponse {
  similar_papers: RecommendationItem[];
  on_this_topic: RecommendationItem[];
  from_these_authors: RecommendationItem[];
}
